src/main/java/org/opensearch/neuralsearch/search/collector/HybridCollapsingTopDocsCollector.java

```diff
@@ -0,0 +1,316 @@
+            return topDocsList;
+        }
+        // Get num sub queries, there is probably a better way to do this
+        int numSubQueries = 0;
+        for (Map.Entry<T, int[]> entry : collectedHitsPerSubQueryMap.entrySet()) {
+            numSubQueries = entry.getValue().length;
+            break;
+        }
+
+        for (int i = 0; i < numSubQueries; i++) {
+            ArrayList<ScoreDoc> fieldDocs = new ArrayList<>();
+            ArrayList<T> collapseValues = new ArrayList<>();
+            int hitsOnCurrentSubQuery = 0;
+            for (T groupValue : groupQueueMap.keySet()) {
+                FieldValueHitQueue<FieldValueHitQueue.Entry> priorityQueue = groupQueueMap.get(groupValue)[i];
+                final int n = priorityQueue.getComparators().length;
+
+                for (int j = 0; j < numHits; j++) {
+                    if (priorityQueue.size() > 0) {
+                        FieldValueHitQueue.Entry queueEntry = priorityQueue.pop();
+
+                        final Object[] fields = new Object[n];
+                        for (int k = 0; k < n; ++k) {
+                            fields[k] = priorityQueue.getComparators()[k].value(queueEntry.slot);
+                        }
+                        fieldDocs.add(new FieldDoc(queueEntry.doc, queueEntry.score, fields));
+                        collapseValues.add(groupValue instanceof BytesRef ? (T) BytesRef.deepCopyOf((BytesRef) groupValue) : groupValue);
+                    } else {
+                        // Break if queue is empty
+                        break;
+                    }
+                }
+                hitsOnCurrentSubQuery += collectedHitsPerSubQueryMap.get(groupValue)[i];
+            }
+            topDocsList.add(
+                new CollapseTopFieldDocs(
+                    collapseField,
+                    new TotalHits(hitsOnCurrentSubQuery, totalHitsRelation),
+                    fieldDocs.toArray(new FieldDoc[0]),
+                    sort.getSort(),
+                    collapseValues.toArray(new Object[0])
+                )
+            );
+        }
+        return topDocsList;
+    }
+
+    @Override
+    public int getTotalHits() {
+        return totalHitCount;
+    }
+
+    @Override
+    public float getMaxScore() {
+        return maxScore;
+    }
+
+    @Override
+    public ScoreMode scoreMode() {
+        return this.needsScores ? ScoreMode.COMPLETE : ScoreMode.COMPLETE_NO_SCORES;
+    }
+
+    @Override
+    public LeafCollector getLeafCollector(LeafReaderContext context) throws IOException {
+        docBase = context.docBase;
+        groupSelector.setNextReader(context);
+        return new HybridLeafCollector() {
+            ConcurrentHashMap<T, Boolean> initializeLeafComparatorsPerSegmentOnceMap;
+
+            {
+                initializeLeafComparatorsPerSegmentOnceMap = new ConcurrentHashMap<>();
+            }
+
+            @Override
+            public void collect(int doc) throws IOException {
```

Comment:
Break down this method code by using Single Responsibility Principle. I am finding it hard time to understand it.

Reply:
Refactored this method